cpp/src/arrow/array/array_test.cc

```diff
@@ -3348,4 +3358,38 @@ TEST(TestSwapEndianArrayData, InvalidLength) {
   }
 }
 
+template <typename T>
+class TestPrimitiveArray : public ::testing::Test {
+ public:
+  void SetUp() { pool_ = default_memory_pool(); }
+
+ protected:
+  MemoryPool* pool_;
+};
+
+TYPED_TEST_SUITE(TestPrimitiveArray, Primitives);
+
+TYPED_TEST(TestPrimitiveArray, IndexOperator) {
+  random::RandomArrayGenerator rng(0x76878);
+
+  auto input_array = rng.ArrayOf(TypeParam::type(), 6, 0.2);
+  auto con_array = checked_pointer_cast<typename TypeParam::ArrayType>(input_array);
+  const auto& carr = *con_array;
+
+  std::vector<util::optional<typename TypeParam::T>> values;
+  for (auto v : carr) {
+    values.push_back(v);
+  }
+
+  for (int64_t i = 0; i < carr.length(); ++i) {
+    auto res = carr[i];
+    if (res) {
+      ASSERT_EQ(values[i].value(), res.value());
```

Comment:
Ok, but this is just testing `operator[]` against the iterator, which are both using exactly the same implementation.

Can you check against the actual array values instead?

Reply:
Exactly my thoughts, hence I tried to use hard coded array as input instead of random number generator, in which case I have to fetch the values from `Array` which indirectly use same routines inside arrow.

Do you mean copy the `int8_t*` buffer and reinterpret it in the test accordingly ?

Reply:
Or you could rely on the existing `carr.Value` method

Reply:
But isn't that the same in a way ? Iterator based fetch is also using `Array::Value` internally. I had that originally and then went with this eventually. I even had the raw copy but that fails with BooleanArray (I found the reason).

Reply:
Ah, you're right. Well, let's not overdo it then. But perhaps you can add type-specific tests for the interval types (using e.g. `ArrayFromJSON` to build up the array)?

Reply:
True, I was trying too much to avoid code duplicate for just those two types. Seems like I can't. Let me add interval type tests separately.